app/ui/browser/model/index.js

```diff
@@ -65,3 +51,20 @@ export class Page extends Immutable.Record({
 Page.PAGE_STATE_LOADING = 'loading';
 Page.PAGE_STATE_LOADED = 'loaded';
 Page.PAGE_STATE_FAILED = 'failed';
+
+/**
+ * The browser window's view of its profile.
+ */
+export const Profile = Immutable.Record({
+  bookmarks: Immutable.Set(),
+  recentBookmarks: Immutable.List(),
+  completions: Immutable.Map(),
+});
+
+/**
+ * Frequently changing UI state.
+ */
+export const UIState = Immutable.Record({
+  // What the user has typed into the location bar, stored by page id
+  userTypedLocation: Immutable.Map(),
```

Comment:
I'm glad you said that because I agree. I decided not to attempt it here because it adds a little complication and I wanted to split it with the smallest amount of extraneous changes, but since you agree I'll work on that in a follow up.